### PR TITLE
Rewrite ModificationExecutor

### PR DESCRIPTION
**Pull request by markuspf**
This pull-request is a rewrite of `ModificationExecutor` and surrounding classes. It provides new Insert, Remove, Replace, Update, and Upsert executors.

This rewrite was triggered by the introduction of `ShadowRows` as part of the Subquery optimisation project currently undertaken by Team Gordo, and us wanting to remove the `SingleBlockFetcher`.

This is a draft, at the very least the follwing things remain to do

 - [x] Handle Statistics
 - [x] Spread some \TRI_ASSERTS` around the code
 - [x] Look at writing some GoogleTests for the new classes
 - [x] Actually hook up the new Modifiers (at the moment they should be swappable with the old ones)
 - [ ] Benchmarks to look outi whether any slowdowns are caused by the new code.
 - [x] Address all TODOs in the code

### CODE REVIEW

File: arangod/Aql/InsertModifier.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+InsertModifierCompletion::InsertModifierCompletion(SimpleModifier<InsertModifierCompletion>& modifier)
+    : _modifier(modifier) {}
+
+InsertModifierCompletion::~InsertModifierCompletion() = default;
+
+Result InsertModifierCompletion::accumulate(InputAqlItemRow& row) {
+  RegisterId const inDocReg = _modifier._infos._input1RegisterId;
+
+  // The document to be UPDATEd
```
Comment (mchacki) [inline arangod/Aql/InsertModifier.cpp]:
```suggestion
  // The document to be INSERTed
```

File: arangod/Aql/ModificationExecutor.h [hunk 2]
```diff
@@ -183,43 +183,26 @@ struct ModificationExecutorInfos : public ExecutorInfos {
...
 
-template <typename FetcherType>
-struct ModificationExecutorBase {
+template <typename Modifier, typename FetcherType>
+class ModificationExecutor {
+  friend struct Insert;
+  friend struct Remove;
+  friend struct Upsert;
+  friend struct UpdateReplace<Update>;
+  friend struct UpdateReplace<Replace>;
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutor.h]: lol names.

However is ok like this ;)

File: arangod/Aql/ModificationExecutor2.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+  builder.close();
+  return Result{};
+}
+
+template <typename FetcherType, typename ModifierType>
+ModificationExecutor2<FetcherType, ModifierType>::ModificationExecutor2(FetcherType& fetcher,
+                                                                        Infos& infos)
+    : _infos(infos), _fetcher(fetcher), _modifier(infos) {
+  // important for mmfiles
+  // TODO: Why is this important for mmfiles?
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutor2.cpp]: In MMFiles we need to make sure that the data is not moved in memory or collected for this collection as soon as we start writing to it.
This pin makes sure that all of the memory moving is not executed and all Pointers we get from a collection stay correct until we release this pin

File: arangod/Aql/ModificationExecutor2.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+ModificationExecutor2<FetcherType, ModifierType>::produceRows(OutputAqlItemRow& output) {
+  TRI_ASSERT(_infos._trx);
+
+  // TODO: isDBServer case
+  ExecutionState state;
+  ModificationExecutor2::Stats stats;
+
+  _modifier.reset();
+
+  size_t maxOutputs = std::min(output.numRowsLeft(), ExecutionBlock::DefaultBatchSize());
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutor2.cpp]: You can rely on numRowsLeft() here.
The outer code makes sure that output is not larger then DefaultBatchSize()

File: arangod/Aql/ModificationExecutor2.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+  if (!transactResult.ok()) {
+    _modifier.throwTransactErrors();
+  }
+
+  doOutput(output, stats);
+
+  // TODO:
+  // We want to say "has more" if upstream still has rows.
+  // sooo... just return the upstream state?
+  return {ExecutionState::DONE, std::move(stats)};
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutor2.cpp]: You have use the `fetchRow()` command above.
This returns a state.
Here you need to return the last state this fetchRow has returned.

File: arangod/Aql/ModificationExecutor2.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+Result buildKeyDocument(VPackBuilder& builder, std::string const& key,
+                        std::string const& rev, bool ignoreRevision = false);
+};  // namespace ModificationExecutorHelpers
+
+//
+// ModificationExecutor2 is the "base" class for the Insert, Remove, Update,
+// Replace and Upsert executors.
+//
+// The fetcher and modification-specific code is spliced in via a template for
+// performance reasons (TODO: verify that this is true).
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutor2.h]:
```suggestion
// performance reasons
```

Comment (mchacki) [reply arangod/Aql/ModificationExecutor2.h]: It is

File: arangod/Aql/ModificationExecutor2.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+// to be submitted to the transaction, a function transact which submits the
+// currently accumulated batch of updates to the transaction, and an iterator to
+// retrieve the results of the transaction.
+//
+// The currently 5 modifier types are divided into the *simple* modifiers
+// Insert, Remove, Update, and Replace, and the Upsert modifier, which is a mix
+// of Insert and Update/Replace, and hence more complicated. (TODO: Is there any
+// way in which we can redesign the execution of AQL queries that we can just
+// transform an upsert into an update and an insert and get rid of the upsert
+// executor?)
```
Comment (mchacki) [inline arangod/Aql/ModificationExecutor2.h]: Regarding TODO:
Not for the time being.
We do not have alternative execution paths

File: arangod/Aql/SimpleModifier.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+template <typename ModifierCompletion, typename Enable>
+SimpleModifier<ModifierCompletion, Enable>::~SimpleModifier() = default;
+
+template <typename ModifierCompletion, typename Enable>
+void SimpleModifier<ModifierCompletion, Enable>::reset() {
+  _accumulator.clear();
+  _operations.clear();
+  _results = OperationResult{};
+  _accumulator.openArray();
```
Comment (ObiWahn) [inline arangod/Aql/SimpleModifier.cpp]: Please do not open and close in different places!

**Comment by goedderz** [comment]
Good idea @ObiWahn.

**Comment by markuspf** [comment]
> Please use the RAII idiom when opening and closing builders. It is really hard to reason about when they are opened and closed in arbitrary locations.

Note that this is not quite as easy as one might like: Scoping with a guard will not work, because the accumulator needs to be able to stay alive across multiple calls to `produceRows`, so we can only do this with object lifetime; Hence this PR now contains a tasty hack that wraps a VPackBuilder inside an object so that we can just create a new accumulator when we need one, and can only add slices (and never close the array which is opened on creation, and closed when the contents of the accumulator are observed. Quantum supremacy.)

**Comment by markuspf** [comment]
This is now being tested here: http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6921/ fingers crossed.

File: arangod/Aql/ModificationExecutorHelpers.cpp [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+using namespace arangodb::aql;
+using namespace arangodb::basics;
+
+Result ModificationExecutorHelpers::getKey(CollectionNameResolver const& resolver,
+                                           AqlValue const& value, std::string& key) {
+  TRI_ASSERT(key.empty());
+
+  // If `value` is a string, this is our _key entry, so we use that.
+  if (value.isString()) {
+    key.assign(value.slice().copyString());
```
Comment (ObiWahn) [inline arangod/Aql/ModificationExecutorHelpers.cpp]: Could be string_view soon https://github.com/arangodb/arangodb/pull/10287 -- no action needed

Comment (ObiWahn) [reply arangod/Aql/ModificationExecutorHelpers.cpp]: Already perfect as `assign` takes `&&`

File: tests/Aql/InsertExecutorTest.cpp [hunk 10]
```diff
@@ -312,12 +312,15 @@ TEST_P(InsertExecutorTestCount, insert_with_key_and_overwrite) {
...
                     " INSERT { _key: TO_STRING(i), value: i } INTO ") +
         collectionName +
-        " OPTIONS { overwrite: true } SORT NEW.value RETURN NEW.value";
+        " OPTIONS { overwrite: true } SORT NEW.value RETURN [OLD.value, NEW.value]";
 
     VPackBuilder builder;
     builder.openArray();
     for (size_t i = 1; i <= nDocs; i++) {
+      builder.openArray();
+      builder.add(VPackSlice::nullSlice());
```
Comment (ObiWahn) [inline tests/Aql/InsertExecutorTest.cpp]: Why do we need the `nullSlice`?

File: arangod/Aql/ModificationExecutorHelpers.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+// Builds an object "{ _key: key, _rev: rev }" if rev is not empty, otherwise
+// "{ _key: key, _rev: null }"
+void buildKeyAndRevDocument(VPackBuilder& builder, std::string const& key,
+                            std::string const& rev);
+
+// Establishes whether a write is necessary. This is only relevant for
+// SmartGraphs in the enterprise edition. Refer to skipForAqlWrite in
+// Enterprise Edition
+bool writeRequired(ModificationExecutorInfos& infos, VPackSlice const& doc,
+                   std::string const& key);
```
[the remark this answers is not in the document]
Comment (markuspf) [reply arangod/Aql/ModificationExecutorHelpers.h]: Yes, if one also corrects the `BoolWrapper` in `ModificationExecutorInfos`. This is probably a site for further later beautification

File: arangod/Aql/ModificationExecutor.cpp [hunk 12]
```diff
@@ -17,118 +17,202 @@
...
+  // This pin makes sure that no memory is moved pointers we get from a collection stay
+  // correct until we release this pin
+  _infos._trx->pinData(this->_infos._aqlCollection->id());
+
+  // TODO: explain this abomination
+  auto* trx = _infos._trx;
+  TRI_ASSERT(trx != nullptr);
+  bool const isDBServer = trx->state()->isDBServer();
+  _infos._producesResults = ProducesResults(
+      _infos._producesResults || (isDBServer && _infos._ignoreDocumentNotFound));
```
[the remark this answers is not in the document]
Comment (markuspf) [reply arangod/Aql/ModificationExecutor.cpp]: Thanks for pointing this out. This also removes the `TODO` above ;)

(I still put an explanation to where the code lives now...)

**Comment by markuspf** [comment]
Thanks to @mchacki @goedderz and @ObiWahn for their reviews, I think I addressed all of the comments now  and started a jenkins PR test  and a performance test. 

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7050/

**Comment by markuspf** [comment]
Tests were blue. Now of course a conflict again...

**Comment by markuspf** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7050/

**Comment by markuspf** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7063/

**Pull request merged**